src/rust/engine/src/nodes.rs

```diff
@@ -1314,22 +1391,47 @@ impl Node for NodeKey {
               }
               message = python_rule_output.message;
               artifacts = python_rule_output.new_artifacts;
-              NodeOutput::Value(python_rule_output.value)
+              NodeOutput::InvocationResult(python_rule_output.value)
             })
             .await
         }
       };
 
-      result = result.map_err(|failure| failure.with_pushed_frame(&failure_name));
+      // Flatten any exceptions into an InvocationResult, but keep everything else unchanged.
+      let (result, was_failure): (NodeOutput, bool) = match result {
+        Ok(result) => match result {
+          NodeOutput::InvocationResult(result) => match result {
```

Comment:
These three nested matches are pretty complex. It would be very good to adjust types to simplify this before landing... possibly via the removal/inlining of either `InvocationResult` or `InternalFailure`, or via any other adjustment you see.